lib/checkstl.cpp

```diff
@@ -2610,10 +2610,25 @@ void CheckStl::useStlAlgorithm()
                 continue;
             const Token *bodyTok = tok->next()->link()->next();
             const Token *splitTok = tok->next()->astOperand2();
-            if (!Token::simpleMatch(splitTok, ":"))
-                continue;
-            const Token *loopVar = splitTok->previous();
-            if (!Token::Match(loopVar, "%var%"))
+            const Token* loopVar{};
+            if (Token::simpleMatch(splitTok, ":")) {
+                loopVar = splitTok->previous();
+                if (!Token::Match(loopVar, "%var%"))
+                    continue;
+            }
+            else if (Token::simpleMatch(splitTok, ";") && Token::simpleMatch(splitTok->astOperand2(), ";")) { // iterator-based loop
+                const Token* comp = splitTok->astOperand2()->astOperand1();
+                loopVar = Token::Match(comp, "%comp%") ? comp->astOperand1() : nullptr;
+                if (!Token::Match(loopVar, "%var%") || !loopVar->valueType() || loopVar->valueType()->type != ValueType::Type::ITERATOR)
+                    continue;
+                const Token* init = splitTok->astOperand1();
+                if (!Token::simpleMatch(init, "=") || !Token::Match(init->astOperand1(), "%varid%", loopVar->varId()))
+                    continue;
+                const Token* inc = splitTok->astOperand2() ? splitTok->astOperand2()->astOperand2() : nullptr;
+                if (!inc || (!Token::Match(inc, "%op% %varid%", loopVar->varId()) && !Token::Match(inc->previous(), "%varid% %op%", loopVar->varId())))
```

Comment:
I think you need to make sure only `++` or `--` is used on the iterator. Also the body of the loop needs to be checked that the iterator isn't further modified.

Reply:
> I think you need to make sure only `++` or `--` is used on the iterator.

E.g. `i += 2` could be realized with `std::accumulate` as well.

> Also the body of the loop needs to be checked that the iterator isn't further modified.

The existing `single***Scope()` functions do that already, I think.

Reply:
> E.g. i += 2 could be realized with std::accumulate as well.

This requires stride iterators which is not available in the standard C++ library.

Reply:
Yeah, I probably thinking of something else. And code like this is ugly and only works with contiguous containers: https://godbolt.org/z/1rGTY6na8